### ISSUE TITLE
Libre Conversion Notification Bug

### ISSUE THREAD

**Issue by kylmcw** [nightscout/Trio#40]
![image](https://github.com/nightscout/Open-iAPS/assets/156543991/25578554-41dd-4732-8566-116b27dbdefe)

Dev branch with live activities and loop_dev_submodules

Notification shows mg/dL when mmol/L selected as glucose value.

Transmitter also shows N/A.

**Comment by pepijnjoost** [comment]
I have noticed the same, but as soon as I disconnected Trio with FSL2 through bluetooth and reconnected it through LibreLink the notifications were in mmol instead of mg/dl.

**Comment by drrree** [comment]
I'm receiving notifications about my blood sugar in mg/dL when I have it set to mmol/L. I even tried to turn off the notification but still get it.

**Comment by dnzxy** [comment]
This is an issue with the Libre module and cannot be solved by the Trio team as of now. Apologies.

**Comment by bjornoleh** [comment]
Tagging @dabear, perhaps something to consider fixing in https://github.com/LoopKit/LibreTransmitter?

**Comment by dabear** [comment]
LibreTransmitterManagerV3 defaults to the  mgdl unit, but will change the glucose value formatter according to events received following the Loopkit DisplayGlucoseUnitObserver protocol unitDidChange method. This is how Loop does that:
```
private var displayGlucoseUnitObservers = WeakSynchronizedSet<DisplayGlucoseUnitObserver>()
//(...)
//updates all observers when there are actual changes to 
func notifyObserversOfDisplayGlucoseUnitChange(to displayGlucoseUnit: HKUnit) {
        self.displayGlucoseUnitObservers.forEach {
            $0.unitDidChange(to: displayGlucoseUnit)
        }
    }

//updates all observers (cgmManagers) immediately when they are created
func addDisplayGlucoseUnitObserver(_ observer: DisplayGlucoseUnitObserver) {
        let queue = DispatchQueue.main
        displayGlucoseUnitObservers.insert(observer, queue: queue)
        queue.async {
            observer.unitDidChange(to: self.displayGlucosePreference.unit)
        }
 }
 
 //(...)
 private extension DeviceDataManager {
    func setupCGM() {
        //(...)

        if let cgmManagerUI = cgmManager as? CGMManagerUI {
            addDisplayGlucoseUnitObserver(cgmManagerUI)
        }
        
final class DeviceDataManager {
 init() {   
 NotificationCenter.default.addObserver(forName: .HealthStorePreferredGlucoseUnitDidChange, object: healthStore, queue: nil) { [weak self] _ in
            guard let self else {
                return
            }

            Task { @MainActor in
                if let unit = await self.healthStore.cachedPreferredUnits(for: .bloodGlucose) {
                    self.displayGlucosePreference.unitDidChange(to: unit)
                    self.notifyObserversOfDisplayGlucoseUnitChange(to: unit)
                }
            }
        }
```

 So this seems like Trio's fault really, not updating the cgmManagers when there is glucose unit changes

**Comment by dabear** [comment]
Trios code:
```
extension CGM {
    struct CGMSettingsView: UIViewControllerRepresentable {
        let cgmManager: CGMManagerUI?
        let bluetoothManager: BluetoothStateManager
        let unit: GlucoseUnits
        weak var completionDelegate: CompletionDelegate?

        func makeUIViewController(context _: UIViewControllerRepresentableContext<CGMSettingsView>) -> UIViewController {
            let displayGlucosePreference: DisplayGlucosePreference
            switch unit {
            case .mgdL:
                displayGlucosePreference = DisplayGlucosePreference(displayGlucoseUnit: .milligramsPerDeciliter)
            case .mmolL:
                displayGlucosePreference = DisplayGlucosePreference(displayGlucoseUnit: .millimolesPerLiter)
            }

            guard let cgmManager = cgmManager else { return UIViewController() }

            var vc = cgmManager.settingsViewController(
                bluetoothProvider: bluetoothManager,
                displayGlucosePreference: displayGlucosePreference,
                colorPalette: .default,
                allowDebugFeatures: false
            )
            // vc.cgmManagerOnboardingDelegate =
            // vc.completionDelegate = self
            vc.completionDelegate = completionDelegate

            return vc
        }
```
It explicitly calls the CGMManager's settings view with displayglucose units derived from Trio's current settings. So that's why it works in the ui

**Comment by dnzxy** [comment]
@dabear We need to ensure that all values that are propagated to Trio from LT are mg/dL and are only parsed for display.
1. Does invoking the unit observer like suggested above change to way values are sent to Trio? 
2. Would be open to looking into this and PRing a fix? I'll re-open the issue and assign you to it, if so 😊

**Comment by dabear** [comment]
Recommendation: On these two events
* CGMManager is created / updated
* Glucose Preferences is changed in Trio

then retrieve the CGMManager as CGMManagerUI , GlucoseUnits from trio settings and apply
```
theCgmManager.unitDidChange(to: unit == .mgdL ? HKUnit.milligramsPerDeciliter ? .millimolesPerLiter)
```

**Comment by dabear** [comment]
You can re-open the ticket and assign it to me. I already fixed it yesterday locally.
During startup:
<img width="1028" alt="image" src="https://github.com/user-attachments/assets/868e4609-6ef8-412b-ab1b-32400992064c">

In the preferences:
<img width="1113" alt="image" src="https://github.com/user-attachments/assets/12027551-49a6-444e-b6c1-f8f030e65e08">
<img width="1113" alt="image" src="https://github.com/user-attachments/assets/01afdeaa-7341-4409-bc3d-698b9c2e299a">

**Comment by dabear** [comment]
what's the target I should make a PR against? Trio Dev branch?
Patch:
```
diff --git a/FreeAPS/Sources/APS/FetchGlucoseManager.swift b/FreeAPS/Sources/APS/FetchGlucoseManager.swift
index 333b4ef7..7b6f44c7 100644
--- a/FreeAPS/Sources/APS/FetchGlucoseManager.swift
+++ b/FreeAPS/Sources/APS/FetchGlucoseManager.swift
@@ -1,5 +1,6 @@
 import Combine
 import Foundation
+import HealthKit
 import LoopKit
 import LoopKitUI
 import SwiftDate
@@ -98,6 +99,14 @@ final class BaseFetchGlucoseManager: FetchGlucoseManager, Injectable {
         settingsManager.settings.uploadGlucose = cgmM.shouldSyncToRemoteService
     }
 
+    private func updateManagerUnits(_ manager: CGMManagerUI?) {
+        let units = settingsManager.settings.units
+        let managerName = cgmManager.map { "\(type(of: $0))" } ?? "nil"
+        let loopkitUnits: HKUnit = units == .mgdL ? .milligramsPerDeciliter : .millimolesPerLiter
+        print("manager: \(managerName) is changing units to: \(loopkitUnits.description) ")
+        manager?.unitDidChange(to: loopkitUnits)
+    }
+
     func updateGlucoseSource(cgmGlucoseSourceType: CGMType, cgmGlucosePluginId: String, newManager: CGMManagerUI?) {
         // if changed, remove all calibrations
         if self.cgmGlucoseSourceType != cgmGlucoseSourceType || self.cgmGlucosePluginId != cgmGlucosePluginId {
@@ -120,6 +129,8 @@ final class BaseFetchGlucoseManager: FetchGlucoseManager, Injectable {
             removeCalibrations()
         } else if self.cgmGlucoseSourceType == .plugin, cgmManager == nil, let rawCGMManager = rawCGMManager {
             cgmManager = cgmManagerFromRawValue(rawCGMManager)
+            updateManagerUnits(cgmManager)
+
         } else {
             saveConfigManager()
         }
@@ -151,7 +162,6 @@ final class BaseFetchGlucoseManager: FetchGlucoseManager, Injectable {
         else {
             return nil
         }
-
         return Manager.init(rawState: rawState)
     }
 
diff --git a/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorDataFlow.swift b/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorDataFlow.swift
index 1bbd54d6..0915350f 100644
--- a/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorDataFlow.swift
+++ b/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorDataFlow.swift
@@ -109,6 +109,7 @@ protocol PreferencesEditorProvider: Provider {
     var preferences: Preferences { get }
     func savePreferences(_ preferences: Preferences)
     func migrateUnits()
+    func updateManagerUnits()
 }
 
 protocol PreferencesSettable: AnyObject {
diff --git a/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorProvider.swift b/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorProvider.swift
index 7e8c63a0..f2a2765e 100644
--- a/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorProvider.swift
+++ b/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorProvider.swift
@@ -1,8 +1,10 @@
 import Foundation
+import HealthKit
 
 extension PreferencesEditor {
     final class Provider: BaseProvider, PreferencesEditorProvider {
         @Injected() private var settingsManager: SettingsManager!
+        @Injected() var fetchGlucoseManager: FetchGlucoseManager!
         private let processQueue = DispatchQueue(label: "PreferencesEditorProvider.processQueue")
 
         var preferences: Preferences {
@@ -17,6 +19,15 @@ extension PreferencesEditor {
             }
         }
 
+        func updateManagerUnits() {
+            var manager = fetchGlucoseManager.cgmManager
+            let managerName = manager.map { "\(type(of: $0))" } ?? "nil"
+            let units = settingsManager.settings.units
+            let loopkitUnits: HKUnit = units == .mgdL ? .milligramsPerDeciliter : .millimolesPerLiter
+            print("manager: \(managerName) is changing units to: \(loopkitUnits.description) ")
+            manager?.unitDidChange(to: loopkitUnits)
+        }
+
         func migrateUnits() {
             migrateTargets()
             migrateISF()
diff --git a/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorStateModel.swift b/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorStateModel.swift
index d28869fb..01175ffb 100644
--- a/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorStateModel.swift
+++ b/FreeAPS/Sources/Modules/PreferencesEditor/PreferencesEditorStateModel.swift
@@ -16,6 +16,7 @@ extension PreferencesEditor {
             subscribeSetting(\.units, on: $unitsIndex.map { $0 == 0 ? GlucoseUnits.mgdL : .mmolL }) {
                 unitsIndex = $0 == .mgdL ? 0 : 1
             } didSet: { [weak self] _ in
+                self?.provider.updateManagerUnits()
                 self?.provider.migrateUnits()
             }
 

```

**Comment by dabear** [comment]
https://github.com/nightscout/Trio/pull/429

**Comment by Sjoerd-Bo3** [comment]
#429 Fixes this

**Comment by Sjoerd-Bo3** [comment]
@dabear Indeed against Dev, and I reopened it and assigned it to you

**Comment by dnzxy** [comment]
@dabear can you please answer this?
> Does invoking the unit observer like suggested above change to way values are sent to Trio?
Will the change you have now PR'd also prompt propagation of mmol/L glucose values, or will they stay (expected behavior on my end) mg/dL?

**Comment by dabear** [comment]
Shouldn't change anything. But since all cgmmanagers get this change you can even test it with dexcom G7 sensors

**Comment by dnzxy** [comment]
Great, thanks ☺️ 
I use G6 + Anubis, no G7s here 😅 but we have plenty testers.

**Comment by dabear** [comment]
Any cgmmanagers including G6..

**Comment by dabear** [comment]
Let me know if anything is needed to get that or merged :) no hurry from my side though

**Comment by Sjoerd-Bo3** [comment]
This is fixed in #429 in Dev and in 0.2.2 Release on Main
